crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -5301,76 +5372,95 @@ impl<'db> TypeInferenceBuilder<'db> {
-                self.infer_type_expression(arguments_slice);
+                self.infer_type_expression_with_qualifiers(arguments_slice);
                 KnownClass::OrderedDict.to_instance(self.db())
             }
             KnownInstanceType::Dict => {
-                self.infer_type_expression(arguments_slice);
+                self.infer_type_expression_with_qualifiers(arguments_slice);
                 KnownClass::Dict.to_instance(self.db())
             }
             KnownInstanceType::List => {
-                self.infer_type_expression(arguments_slice);
+                self.infer_type_expression_with_qualifiers(arguments_slice);
                 KnownClass::List.to_instance(self.db())
             }
             KnownInstanceType::DefaultDict => {
-                self.infer_type_expression(arguments_slice);
+                self.infer_type_expression_with_qualifiers(arguments_slice);
                 KnownClass::DefaultDict.to_instance(self.db())
             }
             KnownInstanceType::Counter => {
-                self.infer_type_expression(arguments_slice);
+                self.infer_type_expression_with_qualifiers(arguments_slice);
                 KnownClass::Counter.to_instance(self.db())
             }
             KnownInstanceType::Set => {
-                self.infer_type_expression(arguments_slice);
+                self.infer_type_expression_with_qualifiers(arguments_slice);
                 KnownClass::Set.to_instance(self.db())
             }
             KnownInstanceType::FrozenSet => {
-                self.infer_type_expression(arguments_slice);
+                self.infer_type_expression_with_qualifiers(arguments_slice);
                 KnownClass::FrozenSet.to_instance(self.db())
             }
             KnownInstanceType::Deque => {
-                self.infer_type_expression(arguments_slice);
+                self.infer_type_expression_with_qualifiers(arguments_slice);
                 KnownClass::Deque.to_instance(self.db())
             }
 
             KnownInstanceType::ReadOnly => {
-                self.infer_type_expression(arguments_slice);
+                self.infer_type_expression_with_qualifiers(arguments_slice);
                 todo_type!("`ReadOnly[]` type qualifier")
             }
             KnownInstanceType::NotRequired => {
-                self.infer_type_expression(arguments_slice);
+                self.infer_type_expression_with_qualifiers(arguments_slice);
                 todo_type!("`NotRequired[]` type qualifier")
             }
-            KnownInstanceType::ClassVar => {
-                let ty = self.infer_type_expression(arguments_slice);
-                ty
-            }
-            KnownInstanceType::Final => {
-                self.infer_type_expression(arguments_slice);
-                todo_type!("`Final[]` type qualifier")
-            }
+            KnownInstanceType::ClassVar | KnownInstanceType::Final => match arguments_slice {
```

Comment:
Should we be passing in some context (maybe as part of `InferContext`?) that tells us whether we're in a class body here? That would allow us to emit a diagnostic here if a definition is annotated with `ClassVar` outside a class body, which is obviously illegal